src/main/scala/diplomacy/AddressDecoder.scala

```diff
@@ -2,111 +2,164 @@
-        x.foreach { a => y.foreach { b =>
-          require (!a.overlaps(b), s"Ports cannot overlap: $a $b")
-        } }
-      }
+      /* Verify the user did not give us an impossible problem. */
+      checkOverlap(nonEmptyPorts)
 
-      val maxBits = log2Ceil(1 + nonEmptyPorts.map(_.map(_.base).max).max)
-      val (bitsToTry, bitsToTake) = (0 until maxBits).map(BigInt(1) << _).partition(b => (givenBits & b) == 0)
-      val partitions = Seq(nonEmptyPorts.map(_.sorted).sorted(portOrder))
+      /** log2 result of the maximum base of [[AddressSet]]s to cover to maximum address. */
+      val maxBits: Int = util.log2Ceil(1 + nonEmptyPorts.map(_.map(_.base).max).max)
+      /** Test each one-hot `base` with [[givenBits]].
+        * If result is 0, we need to try it heuristically.
+        * If result is not 0, we just take it.
+        */
+      val (bitsToTry: Seq[BigInt], bitsToTake: Seq[BigInt]) = (0 until maxBits).map(BigInt(1) << _).partition(b => (givenBits & b) == 0)
+      /** The initial partition. */
+      val partitions: Partitions = Seq(nonEmptyPorts.map(_.sorted).sorted(portOrder))
+      /** Partition [[partitions]] with [[bitsToTake]]. */
       val givenPartitions = bitsToTake.foldLeft(partitions) { (p, b) => partitionPartitions(p, b) }
+      /** Use greedy algorithm to try and select best bits in [[bitsToTry]]. */
       val selected = recurse(givenPartitions, bitsToTry.reverse.toSeq)
+      /** Final bits to partition [[ports]]. */
       val output = selected.reduceLeft(_ | _) | givenBits
-
-      // Modify the AddressSets to allow the new wider match functions
+      /** Modify the AddressSets to allow the new wider match functions. */
       val widePorts = nonEmptyPorts.map { _.map { _.widen(~output) } }
-      // Verify that it remains possible to disambiguate all ports
-      widePorts.combinations(2).foreach { case Seq(x, y) =>
-        x.foreach { a => y.foreach { b =>
-          require (!a.overlaps(b), s"Ports cannot overlap: $a $b")
-        } }
-      }
-
+      /** Verify that it remains possible to disambiguate all ports. */
+      checkOverlap(widePorts)
       output
     }
   }
 
-  // A simpler version that works for a Seq[Int]
+  /** A simpler version that works for a Seq[Int], all mask are set to 0.
+    * @param keys is a a sequence of `base` represented in [[AddressSet]]
+    */
   def apply(keys: Seq[Int]): Int = {
     val ports = keys.map(b => Seq(AddressSet(b, 0)))
     apply(ports).toInt
   }
 
-  // The algorithm has a set of partitions, discriminated by the selected bits.
-  // Each partion has a set of ports, listing all addresses that lead to that port.
-  // Seq[Seq[Seq[AddressSet]]]
-  //         ^^^^^^^^^^^^^^^ set of addresses that are routed out this port
-  //     ^^^ the list of ports
-  // ^^^ cases already distinguished by the selected bits thus far
-  //
-  // Solving this problem is NP-hard, so we use a simple greedy heuristic:
```

Comment:
why did you remove this bit of explanation?

Reply:
I felt like these comments are not in the right place.